enhancements/ingress/global-options-enable-hsts.md

```diff
@@ -0,0 +1,411 @@
+```yaml
+    hstsHeaderPolicy:
+      scope:  all
+      hstsHeader: "max-age=31536000;includeSubDomains;preload"
+```
+
+The `spec.httpHeaders.hstsHeaderPolicy` is formatted and copied to the router environment variable `ROUTER_DEFAULT_HSTS` when the
+`IngressOperator` creates the router.  Another environment variables holds the scope `ROUTER_HSTS_SCOPE`.  Changes are needed in `pkg/operator/controller/ingress/deployment.go`:
+```go
+const (
+	WildcardRouteAdmissionPolicy = "ROUTER_ALLOW_WILDCARD_ROUTES"
+	...
+	RouterHSTSHeaderPolicy       = "ROUTER_DEFAULT_HSTS"
+	RouterHSTSHeaderScope        = "ROUTER_HSTS_SCOPE"
+    ...
+)
+...
+func desiredRouterDeployment(ci *operatorv1.IngressController, ingressControllerImage string, ingressConfig *configv1.Ingress, apiConfig *configv1.APIServer, networkConfig *configv1.Network, proxyNeeded bool) (*appsv1.Deployment, error) {
+...
+    if ci.Spec.HTTPHeaders != nil && ci.Spec.HTTPHeaderPolicy != nil && len(ci.Spec.HSTSHeaderPolicy.MaxAge) != 0 {
+        header := "maxage=" + ci.Spec.HTTPHeaders.HSTSHeaderPolicy.MaxAge
+        if ok := ci.Spec.HTTPHeaders.HSTSHeaderPolicy.IncludeSubdomains; ok != nil && ok {
+        	header = header + ";includeSubdomains"
+        }
+        if  ok = ci.Spec.HTTPHeaders.HSTSHeaderPolicy.Preload; ok != nil && ok {
+        	header = header +  ";preload"
+        }
+        env = append(env, corev1.EnvVar{Name: RouterHSTSHeaderPolicy, Value: header})
+        scope = ci.Spec.HTTPHeaders.HSTSHeaderPolicy.Scope
+        if scope == "limited" { // more details on this later in the EP doc
+            env = append(env, corev1.EnvVar{Name: RouterHSTSHeaderScope, Value: "limited"})
+        } else {
+            env = append(env, corev1.EnvVar{Name: RouterHSTSHeaderScope, Value: "all"})
+        }
+    }
+...
+```
+
+We don't do any other header validation, but if needed it could go in here:
+https://github.com/openshift/cluster-ingress-operator/blob/master/pkg/operator/controller/ingress/controller.go#L441
+
+Accompanying unit and e2e test changes are added to exercise the new header.
+
+With the addition of processing for `ROUTER_DEFAULT_HSTS`, the code template snippet above becomes:
+```gotemplate
+{{- $router_global_hsts := env "ROUTER_DEFAULT_HSTS" "" }}
+...
+{{- if matchValues (print $cfg.TLSTermination) "edge" "reencrypt" }}
+  {{- /* First check the environment to see if the header is set globally */}}
+  {{- if (ne "" $router_global_hsts) }} 
+    http-response set-header Strict-Transport-Security {{$router_global_hsts}}
+  {{- end }}{{/* global_hsts */}}
+  {{- /* Override if the annotation exists.  Set-header replaces any existing header */ }}
+  {{- with $hsts := firstMatch $hstsPattern (index $cfg.Annotations "haproxy.router.openshift.io/hsts_header") }}
+    http-response set-header Strict-Transport-Security {{$hsts}}
+  {{- end }}{{/* hsts header override per-route */}}
```

Comment:
It would be preferable to avoid writing two `http-response set-header Strict-Transport-Security` lines if both the global setting and the annotation are set:
```suggestion
  {{- with $hsts := firstMatch $hstsPattern (index $cfg.Annotations "haproxy.router.openshift.io/hsts_header") $router_global_hsts }}
    http-response set-header Strict-Transport-Security {{$hsts}}
  {{- end }}
```

Reply:
I have updated the design to use a helper function for the HAproxy template, `hstsHeaderValue`.  Needed to revert the `enums` for scope, back to `string consts.`